ops/main.py

```diff
@@ -356,9 +325,255 @@ def _should_use_controller_storage(db_path: Path, meta: CharmMeta) -> bool:
         return False
 
 
+@dataclasses.dataclass(frozen=True)
+class _CharmSpec:
```

Comment:
I don't really understand the utility of the `_CharmSpec` class. The `autoload` method overlaps a lot with `CharmMeta.from_yaml` (with the addition of loading `config.yaml`, but that could easily be an addition to `CharmMeta`).

It seems like the PR would be simpler if `Ops` took a charm type as an `__init__` argument, and the `_get_charm_dir` method could be in there, along with one to load the metadata (some of the minor changes like checking if the metadata exists are nice and could stay).

Am I maybe missing use cases for the _CharmSpec class? If so, could you outline/link to those? Maybe in the PR description rather than specifically in this thread for better visibility?

Reply:
Added some more thoughts on the PR description, but TLDR: encapsulation and ease of mocking. If you look at how I'm using it in scenario, it should become more clear what the intended use case is.